Fix MPI hanging bug.

mgjarrett opened the pull request:
## Description

Making the fix from #688 more robust. There are some situations where `MPI_SIZE` can be greater than 1, but `MPI_DISTRIBUTABLE` is `False` and only the rank 0 process has called `cleanPath()`. This code prevents the code from getting to an infinite MPI barrier in that situation.

---

## Checklist

- [x] The code is understandable and maintainable to people beyond the author.
- [x] Tests have been added/updated to verify that the new or changed code works.
- [x] There is no commented out code in this PR.
- [x] The commit message follows [good practices](https://terrapower.github.io/armi/developer/tooling.html).
- [x] All docstrings are still up-to-date with these changes.

If user exposed functionality was added/changed: N/A

- [ ] Documentation added/updated in the `doc` folder.
- [ ] New or updated dependencies have been added to `setup.py`.

john-science commented on armi/utils/pathTools.py:
```diff
@@ -240,7 +240,7 @@ def cleanPath(path, mpiRank=0):
     valid = False
     if not os.path.exists(path):
-        if context.MPI_SIZE > 1:
+        if context.MPI_SIZE > 1 and context.MPI_DISTRIBUTABLE and barrier:
```
For computational efficiency, can we write this as:

> if barrier and context.MPI_SIZE > 1 and context.MPI_DISTRIBUTABLE:

The idea here is there is no real cost to checking the value of `barrier`, but there is an incurred cost to pulling up the `context` object and checking the namespace inside it.

This is not super important, but it's also easy to fix.

john-science commented on armi/utils/pathTools.py:
```diff
@@ -283,7 +283,7 @@ def cleanPath(path, mpiRank=0):
 
     # Potentially, wait for all the processes to catch up.
-    if context.MPI_SIZE > 1:
+    if context.MPI_SIZE > 1 and context.MPI_DISTRIBUTABLE and barrier:
```
Same as my other comment on the ordering of this multi-boolean check.

mgjarrett commented:
I moved the MPI barrier outside of `cleanPath`, and I adjusted the logic for calling the barrier slightly. I think this is the right logic to call MPI barrier only when processes are running in parallel.

mgjarrett commented:
It's difficult to meaningfully cover many of the places where `cleanPath()` and `context.waitAll()` are called because they happen when cleaning up folders from ARMI runs, but those folders don't get generated when running the unit tests.

mgjarrett commented:
Still having trouble with downstream applications... don't merge this yet. I'm stuck in a position where the code that passes unit test fails downstream in-use tests, and vice versa.

john-science commented:
> Still having trouble with downstream applications... don't merge this yet. I'm stuck in a position where the code that passes unit test fails downstream in-use tests, and vice versa.

Tell me when you're good, man.

opotowsky commented:
@jstilley I just chatted with @mgjarrett and he says it's ready!
